tools/util/half.h

```diff
@@ -209,7 +215,7 @@ std::string lexical_cast<std::string>(cutlass::half_t const& arg);
 
 #define HLF_MANT_DIG 10
 
-namespace std {
+namespace cutlass {
```

Comment:
Can you elaborate the reasoning behind this change? It was a deliberate choice to place the code here under std

Reply:
Good question. It's a leftover from an early attempt to make cutlass compile with clang.
It appears to be unnecessary now. I'll remove it.

Reply:
It turns out the issue is a bit more complicated. Could you elaborate on the reasons for using `namespace std` here?

Technically, IIUIC, you are not allowed to extend std:: namespace this way -- it's undefined behavior that happens to work with libstdc++. It fails with libc++, which is what I was running into.
Consider this reduced case: https://godbolt.org/z/29jOBD -- when `namespace std` is used, it compiles with libstdc++ with both nvcc and clang. However, it fails with clang+libc++.

Using `namespace cutlass` here allows the standard library to find the functions via ADL and works with both libc++ and libstdc++.  

@timshen91 -- Tim, what's the official stand on extending standard c++ library this way?

Reply:
According to the C++ standard [namespace.std]:
* Unless otherwise specified, the behavior of a C++ program is undefined if it adds declarations or definitions to namespace std or to a namespace within namespace std.
* Unless explicitly prohibited, a program may add a template specialization for any standard library class template to namespace std provided that (a) the added declaration depends on at least one program-defined type and (b) the specialization meets the standard library requirements for the original template.

That basically says that defining std::isnan/isinf/copysign/etc for half_t causes undefined behavior, but specializing numeric_limits for half_t is fine.

For the practical side, the standard has many reasons to forbid such arbitrary cutomization. For example, libc++ implements everything in `namespace std { inline namespace __1 { ... } }`. In that case, half_t version of isnan doesn't even overload with std::__1::isnan.

The common way of injecting user customization into standard library functions is through ADL (https://en.cppreference.com/w/cpp/language/adl), which requires the customization (isnan) to be in the same namespace as the type (half_t), hence this change.

Reply:
Thank you for explaining this, Tim and Artem. I'm definitely convinced.  Your proposed solution of defining these functions in namespace `cutlass` sounds like the correct approach.